data/tutorials/ds_05_seq.md

```diff
@@ -0,0 +1,317 @@
+---
+id: sequences
+title: Sequences
+description: >
+  Learn about an OCaml's most-used, built-in data types
+category: "data-structures"
+date: 2023-01-12T09:00:00-01:00
+---
+
+# Sequences
+
+## Prerequisites
+
+| Concept | Status | Documentation | Reference |
+|---|---|---|---|
+| Basic types | Mandatory | | |
+| Functions | Mandatory | | |
+| Lists   | Mandatory |   |   |
+| Options | Recommended |   |   |
+| Arrays  | Nice to have |   |   |
+
+## Introduction
+
+Sequences look a lot like lists. However from a pragmatic perspective, one
+should imagine they may be infinite. That's the key intuition to understanding
+and using sequences.
+
+One way to look at a value of type `'a Seq.t` is to consider it as an icicle, a
+frozen stream of data. To understand this analogy, consider how sequences are
+defined in the standard library:
+```ocaml
+type 'a node =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = unit -> 'a node
+```
+This is the mutually recursive definition of two types; `Seq.node` which is
+almost the same as `list`:
+```ocaml
+type 'a list =
+  | []
+  | (::) of 'a * 'a list
+```
+and `Seq.t` which is merely a type alias for `unit -> 'a Seq.node`. The whole
+point of this definition is the type of the second argument of `Seq.Cons`, which
+is a function returning a sequence while its `list` counterpart returns a list. Let's
+compare the constructors of `list` and `Seq.node`:
+1. Empty lists and sequences are defined the same way, a constructor without any
+   parameter: `Seq.Nil` and `[]`.
+1. Non-empty lists and sequences are both pairs whose former member is a piece
+   of data;
+1. but the latter member, in lists, is a `list` too, while in sequences, it is a
+   function returning a `Seq.node`.
+
+A value of type `Seq.t` is “frozen” because the data it contains isn't
+immediately available, a `unit` value has to be supplied to recover it, and
+that's “unfreezing”. However, unfreezing only gives access to the tip of the
+icicle, since the second argument of `Seq.Cons` is a function too.
```

Comment:
```suggestion
A value of type `Seq.t` is “frozen” because the data it contains isn't
immediately available. A `unit` value has to be supplied to recover it, which is called “unfreezing.” However, unfreezing only gives access to the tip of the
icicle, since the second argument of `Seq.Cons` is a function too.
```